domains/transfer/Transfer.API/API/Claiming/Automation/ClaimService.cs

```diff
@@ -0,0 +1,81 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using API.Claiming.Api.Repositories;
+using API.Claiming.Automation.Services;
+using Microsoft.Extensions.Logging;
+using ProjectOrigin.WalletSystem.V1;
+
+namespace API.Claiming.Automation;
+
+public class ClaimService : IClaimService
+{
+    private readonly ILogger<ClaimService> logger;
+    private readonly IClaimRepository claimRepository;
+    private readonly IProjectOriginWalletService walletService;
+
+    public ClaimService(ILogger<ClaimService> logger, IClaimRepository claimRepository, IProjectOriginWalletService walletService)
+    {
+        this.logger = logger;
+        this.claimRepository = claimRepository;
+        this.walletService = walletService;
+    }
+
+    public async Task Run(CancellationToken stoppingToken)
+    {
+        while (!stoppingToken.IsCancellationRequested)
+        {
+            logger.LogInformation("ClaimService running at: {time}", DateTimeOffset.Now);
+            try
+            {
+                var claimSubjects = await claimRepository.GetClaimSubjects();
+                foreach (var subjectId in claimSubjects.Select(x => x.SubjectId).Distinct())
+                {
+                    var certificates = await walletService.GetGranularCertificates(subjectId);
+
+                    var certificatesGrouped = certificates.GroupBy(x => new { x.GridArea, x.Start, x.End });
+
+                    foreach (var cert in certificatesGrouped)
+                    {
+                        var productionCerts = cert.Where(x => x.Type == GranularCertificateType.Production).ToList();
+                        var consumptionCerts = cert.Where(x => x.Type == GranularCertificateType.Consumption).ToList();
```

Comment:
The order of the certificates may be fixed by the wallet (e.g. they are ordered by issuance date), which with this claiming/matching algorithm will introduce a bias to which certificates are matched/claimed first.
I will suggest that the certificates are shuffled/ordered randomly at some point to ensure that this algorithm is not biased.

Reply:
Why do you suggest that we actively do something to shuffle it? What do we gain?

Reply:
We would ensure that it is actually shuffled - this means that there is no bias in the algorithm.

Reply:
Why do you suggest that we do not shuffle it? Why do you want to base the matching on the order of the certificates returned by the Wallet?

Reply:
Because we actively have to shuffle it - this looks cleaner. Isn't it ok with bias? Then we know what certificates it takes first